staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -74,6 +74,26 @@ type ValidatingAdmissionPolicy struct {
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 	// Specification of the desired behavior of the ValidatingAdmissionPolicy.
 	Spec ValidatingAdmissionPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Status is the status of the ValidatingAdmissionPolicy, including warnings that are useful to determine if the policy
+	// behaves in the expected way.
+	// +optional
+	Status ValidatingAdmissionPolicyStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
when there are multiple kube-apiserver instances, and multiple aggregated non-kube-apiserver API server instances all running the CEL / ValidatingAdmissionPolicy admission plugin/controller, will they all be trying to write to this field?

Reply:
Short answer is yes. Status is part of the object, even if it uses different REST/storage it is still the same object. concurrent update/apply is possible.
Long answer is that the same controller running multiple instances between controllers is not part of the design. I remember creating a Leader Migration feature specifically to avoid running multiple instances of the same controller.
By introducing a Status, the ValidatingAdmissionPolicy object becomes "stateful" and thus have this concurrent write problem. I think if that turns out to be a problem in real world we should work on a more detailed design, in the KEP.

Reply:
all kube-apiserver instances writing is not ideal, but not different than CRDs (though the cost of this seems higher than the CRD status writer, so it's not great to do that work for nothing). 

Aggregated apiserver instances needing/having permission to write status for these is a bigger issue, in my opinion, but saying the kube-apiserver instance of the admission controller in the generic apiserver library is the only one that should write to status seems really weird as well.

Reply:
Re multiple kube-apiservers, [the CRD non-structural-schema controller uses generation](https://github.com/kubernetes/kubernetes/blob/f461527491effcdbe23814b7481f46723cddda9e/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/nonstructuralschema/nonstructuralschema_controller.go#L141-L147). I had [proposed solution in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3488-cel-admission-control/README.md#writing-to-status) based on that, to use here.

The aggregated apiserver case is more troubling.. Given our timeline and that we're still in alpha, I'm in favor of documenting that we only produce these compiler status warnings for kube-apiserver owned types. I realize that's not optimal, but it will give us experience with this feature before we commit to building it out further.

Reply:
> I'm in favor of documenting that we only produce these compiler status warnings for kube-apiserver owned types
That would be an easy change. Swap the schema resolver to only the built-in should do the trick.

Reply:
If multiple controller does their own job. Would you think it would be better for each aspect to track its own generation?
For example:
```yaml
status:
    typeChecking:
        generation: 1
        expressionWarnings:
        - expression: "foo.foo > 1"
          warning: "warning here"
```

Reply:
As an update. Here is the current solution.

1. `ObservedGeneration` is added to `ValidatingAdmissionPolicyStatus` (see below), the status reconciles independently from the Spec
2.  Type checking only happens when the observed generation is older than the object. This solves the race where an API server overwrites the Status with one that matches a previous generation.
3. The type checking algorithm is deterministic. No matter which instances of the admission controller does the type checking, the resulting status would be the same. We will require any future addition to the Status object to have the same property.

Please let me know if the current solution addresses your concern. Thank you.

Reply:
> As an update. Here is the current solution.
> 
> 1. `ObservedGeneration` is added to `ValidatingAdmissionPolicyStatus` (see below), the status reconciles independently from the Spec

:+1: 

> 2. Type checking only happens when the observed generation is older than the object. This solves the race where an API server overwrites the Status with one that matches a previous generation.

:+1: 

> 3. The type checking algorithm is deterministic. No matter which instances of the admission controller does the type checking, the resulting status would be the same. We will require any future addition to the Status object to have the same property.

Type checking for a specific a specific combination `ValidatingAdmissionPolicy{metadata.generation=X}, ParamKind{metadata.generation=Y}, MatchKind{metadata.generation=Z}` is deterministic. 

If we only have an `ObservedGeneration` for the `ValidatingAdmissionPolicy`, then I don't see how the controller would know that a change to the ParamKind should trigger a type check.

I would expect a full solution to deterministically output the same type warnings based on the current state of the 3 resources, regardless of the order of operations that were used to reach that state. For example, I would expect that creating a `ValidatingAdmissionPolicy` that has `params.x` in an expression to produce a status warning if the `ParamKind` does not have a field `x`. It should matter if the `ParamKind` is created without the field or if it is created with  the field and the field is then later deleted.

> 
> Please let me know if the current solution addresses your concern. Thank you.

Reply:
As an update. CRD checking is disabled in this PR. We will need a separate PR to provide a solution of the CRD mutation problem. For now, the concurrent write problem is solved as long as we deal with only built-in types.